### PR TITLE
[Don't merge yet][T5, Bart] Allow t5 torch trace

### LINKED ISSUE #5647: T5 TorchScript (Trace) Conversion
# ❓ Questions & Help

How can we correctly set inputs for t5 TorchScript? 

## Details

```
from transformers import *

tokens_tensor = torch.ones(1, 10, dtype=torch.long)
model = T5Model.from_pretrained("t5-small", torchscript=True)
model.eval()
scripted_model = torch.jit.trace(model, (tokens_tensor, tokens_tensor))
```

Error:
```
ValueError: You have to specify either decoder_input_ids or decoder_inputs_embeds
```

### PR DESCRIPTION
**Pull request by patrickvonplaten**
This PR would fix #5647 .

It's not a great solution IMO though. 
The problem with torch script is that one **cannot** pass keyword arguments, but has to pass positional arguments and it is not possible to pass `None` because every input is required to be a tensor.
Because T5 requires both `input_ids` and `decoder_input_ids`, the two arguments should arguably be placed as the first two arguments.

There might be use cases though, where the same error would occur, which we could not save then, *e.g.* one wants to input `input_embeds`.

Maybe @LysandreJik @sgugger have a better idea.

### CODE REVIEW

**Comment by sgugger** [comment]
Looks innocuous to me, is there some test this allows us to enable for jit and T5?

File: tests/test_modeling_common.py
Comment (patrickvonplaten) [inline tests/test_modeling_common.py:248]: Quite hacky here, but I didn't see another way....

File: tests/test_modeling_t5.py
Comment (patrickvonplaten) [inline tests/test_modeling_t5.py:285]: turn test on

File: tests/test_modeling_common.py
Comment (sgugger) [reply tests/test_modeling_common.py:248]: Isn't it more general as `is_encoder_decoder`?

Comment (patrickvonplaten) [reply tests/test_modeling_common.py:248]: Think `bart` can run without requiring the `decoder_input_ids` @sshleifer...but I guess it would be cleaner to call it `encoder_decoder` here...we will have to slightly change Bart then.

Comment (patrickvonplaten) [reply tests/test_modeling_common.py:248]: If I change to `is_encoder_decoder=True`, and do the corresponding changes for Bart, Bart hits an assert: 

```
assert attn_output.size() == (bsz * self.num_heads, tgt_len, self.head_dim)
```

I guess `decoder_input_ids` should not be the same length as `input_ids` in Bart @sshleifer ?

Comment (patrickvonplaten) [reply tests/test_modeling_common.py:248]: I'll push the changes -> let's see if we want to change Bart accordingly or revert to the T5 hack...

Comment (sshleifer) [reply tests/test_modeling_common.py]: I wrote that assert and should have written an error message, but I can't understand the issue without looking more closely. Does it matter that the second arg to `BartForConditionalGeneration.forward` is `attention_mask`?

**Review by sshleifer — APPROVED** [review]

**Review by sshleifer — APPROVED** [review]

File: tests/test_modeling_common.py
Comment (sshleifer) [reply tests/test_modeling_common.py]: Its outdated, seems like you figured it out! I will add a message to my assert.

**Comment by patrickvonplaten** [comment]
After digging a bit deeper why Bart tests fail, I think the reason is that the Bart cache/`past_key_value` data structure is not compatible with torchscript. For now the Bart tests pass because returning the cache/`past_key_vaule` is disabled - see https://github.com/huggingface/transformers/blob/ac001c48b8df1f5aadcc8cf2c71d7c1116c05250/tests/test_modeling_common.py#L252. 

A bug was filed for this problem: https://github.com/huggingface/transformers/issues/6348.

**Comment by patrickvonplaten** [comment]
PR should be good for merge IMO. @LysandreJik @sshleifer @sgugger - would be great if you can take a quick second look.

File: tests/test_modeling_common.py
Comment (sshleifer) [inline tests/test_modeling_common.py:263]: (out of scope) 
Why is the try/except -> `self.fail` pattern useful?

Without try/except you get an error that traces back to a line with the word `save` in it.
I just looked at `self.fail` and it turns things into AssertionErrors, which seems like strictly less info.

**Review by sshleifer — APPROVED** [review]
LGTM. I'll do my part after this merges!

I'm a little concerned that jit cares about arg ordering, but python doesn't care about kwarg ordering, so there might be some very confusing bugs in the future. Don't have a great idea of how to fix + maintain the same API.

**Review by sgugger — APPROVED** [review]
LGTM, great work!

**Comment by patrickvonplaten** [comment]
Putting this on hold for now as it introduces a breaking change.

**Comment by misrasaurabh1** [comment]
Any updates on this?

**Comment by patrickvonplaten** [comment]
The problem is that it breaks backwards compatibility in a sense that the positional arguments of Bart and T5 are changed. At the moment this is the only option to make torch tracing work for Bart and T5 though...there might be a possiblity to trace a wrapper around the model though

**Pull request closed without merging**
